### ISSUE TITLE
Give webcam focus to the current speaker

### ISSUE THREAD

**Issue by ffdixon** [bigbluebutton/bigbluebutton#8388]
[ I'm pretty sure there is an issue opened for this already, but unable to locate it so creating a new one. ]

When 3 (or more) webcams are shared, we currently enable a user to manually focus on a particular webcam (make it larger).

Extend the speaker detection so the webcam for the current speaker has focus.

**Comment by prlanzarin** [comment]
Might not want to use the speaker/talking indicator as it cycles too easily and might make the focus change too distracting. FreeSWITCH has another event called `video-floor-change` for a speaker floor that already takes energy levels and transition periods into account and is much better for this kind of stuff.

**Comment by capilkey** [comment]
The switching being too frequent was the problem that I ran into when I tried to do this in the Flash client. It's very easy to make a talker's webcam have focus, but it's more tricky to make it function in a way that works well.

**Comment by prlanzarin** [comment]
The video-floor event is pretty reliable (and configurable). The algorithm FreeSWITCH uses has seen some use on my part for a scenario similar to this one and it should be a good fit. We need to parse it in fsesl and propagate to the client (or use SFU to do that since it already fetches and propagates the event as well). Either way I think it's doable with `video-floor-change`, but it needs to be optional and behind a flag.

**Comment by ffdixon** [comment]
I think the function to give focus would need a bit of impedance to not switch too quickly, but otherwise it should be possible to create.

**Comment by capilkey** [comment]
The energy level is actually a pretty important part because there's a big difference between a talker that just has some background noise and someone with higher levels of noise. The client doesn't have that energy info so all talkers are judged the same.

I kind of like the idea of the video-floor-change. Does it work if we aren't sending the videos through FS?

**Comment by prlanzarin** [comment]
Yes, it does. The naming is not precise, but it's used in mod_conference to handle the video mixing layout and it's fired for every audio user in the bridge. I have it implemented both in SFU (https://github.com/bigbluebutton/bbb-webrtc-sfu/blob/master/lib/mcs-core/lib/adapters/freeswitch/esl-wrapper.js#L348-L355) and akka-fsesl/apps (through the standard Redis part) in a fork of mine. I'll lift it to upstream and open a PR later on.

**Comment by SteveClement** [comment]
In a round table type of conference this is ideal.
AI to the resque?

Perhaps this might be a moderator feature.

**Comment by mtsonline** [comment]
Hi,

great that you are working on that, our users would very much appreciate this focus!

as it is a bit related to teh "is speaking" message and "energy-level" used for it:
#7007 

could we also use the same indication (from the video floor) or a second "display-energy-level" for the "is speaking" messages?
The problem there is, that many users now are setting energy-level to 0 because it breaks the audio quality and causes crackling noises.
And even if you use energy-level, the quick changing "is speaking" is not very useful - so this should be the same as for the video I think.

**Comment by pikatowotsa** [comment]
It would really be nice to have this feature allowing the moderator to set focus on the users cam.

**Comment by agussang** [comment]
> It would really be nice to have this feature allowing the moderator to set focus on the users cam.

Any update for this feature? i need feature for moderator set camera focused to attendance, or showing camera attendance to other attendance by moderator on locked attendance, or giving priviledges to presenter for attendance but presenter can't view cam attendancr

**Comment by pbdco** [comment]
Any news on this?

**Comment by mtsonline** [comment]
I think we should wait for the upcoming 2.3 release, there will be a lot of changes and we will see which feature requests could be implimented.  :-)

**Comment by miztaka** [comment]
:+1:

**Comment by GremL1N** [comment]
Maybe for beginning we can implement focusing by moderator like presentation update on changes? I don't know JS and Meteor maybe someone can give me advise and i'll try
https://docs.bigbluebutton.org/dev/dev23.html were nothing about focus in feature list.

**Comment by einhirn** [comment]
Mentioning #13869 because there the "floor holder" flag also triggers a change concerning videos. Maybe, code wise, it's only a small step from there? Although this is more of a UI/layout manager thing than Video related, right?

Re: "configurable and behind a flag": I guess this behavior shouldn't be configured on the server only, but (also) in UI for each user individually? And that option should also be included in layout push. (Which will be overhauled in 2.5 anyway so far as I understand #14146).

**Comment by prlanzarin** [comment]
> Mentioning #13869 because there the "floor holder" flag also triggers a change concerning videos. Maybe, code wise, it's only a small step from there? Although this is more of a UI/layout manager thing than Video related, right?

Code-wise, yeah, it's straightforward now. The thing is UI/UX. 
I think [a comment of mine in another issue](https://github.com/bigbluebutton/bigbluebutton/pull/13835#issuecomment-1015538935) covers my concerns about this currently (just replace the places I'm talking about `pin` by `floor holder`. Especially this part:
```
Can't be a simple "focus on pin" action because focus just looks weird/is worse right now under some circumstances and we still don't have something similar to a "spotlight" grid like Zoom or Google Meet do.
```

**Comment by einhirn** [comment]
Under what circumstances does focus look weird? And would that weirdness be tolerable, if each user can decide whether to enable/use that feature?

**Comment by easy-bios** [comment]
In #14340 I suggested an other solution to have a big prominent webcam set by moderatior or presentator:
![image](https://user-images.githubusercontent.com/42430994/154075297-08d1b777-faa5-4514-8810-15c82e68a934.png)

Thist would be very nice for presentations without screenshare etc. but only via webcam.

**Comment by PavelZ-creator** [comment]
Any news on this, please? I believe that absence of a layout focusing on the speaker such as the one suggested by @easy-bios  is one of the biggest obstacles preventing BBB from becoming more popular choice among video conference programs.

**Comment by prlanzarin** [comment]
No news yet, but I'm linking this to https://github.com/bigbluebutton/bigbluebutton/issues/13844 so that this is factored into the grid UI/UX re-thinking.

We didn't forget about this issue. When it's actually assigned you'll see a milestone assignment/tagging and further comments.
